### PR TITLE
Clarify dependencies: required vs optional, and build vs runtime.

### PR DESCRIPTION
**Pull request by terrycloth**
## Description

A first pass at updating the dependency documentation.

Fixes issue #2062.

## TODOs:

- [ ] Double check these changes for accuracy, make sure I understood the discussion and existing documentation correctly.
- [ ] Maybe scan fish functions for more hard or optional dependencies.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -13,17 +13,32 @@ Detailed user documentation is available by running `help` within fish, and also
...
+* a curses implementation, such as ncurses.
+* gettext, if you need the localizations.
+* PCRE2, due to the regular expression support contained in the `string` builtin. A copy is included with the source code, and will be used automatically if it does not already exist on your system.
+* a number of common UNIX utilities: coreutils (at least ls, seq, rm, mktemp), hostname, grep, awk, sed, and getopt.
+* bc, the "basic calculator" program.
 
-Building the documentation requires Doxygen 1.8.7 or newer.
+Optional functionality should degrade gracefully. If fish ever threw an error about missing dependencies for an optional feature, it would be a bug in fish and should be reported on our GitHub. Optional features and dependencies include:
+
+* fish's builtin `test` syntax, which requires `diff` and `expect`. You don't need these, though, if you don't want to use our tests.
```
Comment (faho) [inline README.md]: This isn't about the `test` builtin (as in `test 5 -eq 8`), but about our test suite.

File: README.md [hunk 2]
```diff
@@ -13,17 +13,32 @@ Detailed user documentation is available by running `help` within fish, and also
...
+* dynamically generating usage tips for builtin functions, which requires man, apropos, nroff, and ul.
+* manual page parsing for auto-completion for external commands, which also requires the above man tools, plus Python 2.6+ or Python 3.3+. Python 2 will also need the backports.lzma package; Python 3.3+ should include the required module by default.
+* the web configuration tool, which requires Python 2.6+ or 3.3+.
+* the fish_clipboard_* functions (bound to \cv and \cx), which require xsel or pbcopy/pbpaste.
+
+fish can also show VCS information in the prompt, when the current working directory is in a VCS-tracked project. This is one of the weakest dependencies, though: if you don't use git, you don't need git information in the prompt, and if you do use git, you already have it.
```
Comment (faho) [inline README.md]: I'd mention git, hg and svn.

File: README.md [hunk 3]
```diff
@@ -13,17 +13,32 @@ Detailed user documentation is available by running `help` within fish, and also
...
 
-fish requires gettext for translation support.
+* a curses implementation, such as ncurses.
+* gettext, if you need the localizations.
+* PCRE2, due to the regular expression support contained in the `string` builtin. A copy is included with the source code, and will be used automatically if it does not already exist on your system.
+* a number of common UNIX utilities: coreutils (at least ls, seq, rm, mktemp), hostname, grep, awk, sed, and getopt.
+* bc, the "basic calculator" program.
 
-Building the documentation requires Doxygen 1.8.7 or newer.
+Optional functionality should degrade gracefully. If fish ever threw an error about missing dependencies for an optional feature, it would be a bug in fish and should be reported on our GitHub. Optional features and dependencies include:
```
Comment (faho) [inline README.md]: Note: What I meant by that is if you run e.g. `fish_update_completions` and it says "Syntax error in the frobulator on line 252525, dumping core....", that would be a bug.

If you try to run that and it says "Please install python", that's only fair. (Currently it prints nothing)

It should not print a message on first start (when it's trying to generate completions).

**Comment by faho** [comment]
One more thing: I'd add a note saying that it is to be expected that completion scripts call the tool they are completing, and anything usually shipped with it.

File: README.md (hunk 3 above)
Comment (terrycloth) [inline README.md]: Good point, thanks.

File: README.md (hunk 1 above)
Comment (terrycloth) [inline README.md]: Ah, that makes sense. I think the original text was a little ambiguous, but I'll clarify it when I get a moment later today.

Comment (terrycloth) [inline README.md]: Actually, if it's only for testing fish itself during development, maybe the test dependencies should go in the CONTRIBUTING documentation instead.

**Comment by terrycloth** [comment]
I also noticed the .editorconfig file specifies a max line length of 100 characters, but all the Markdown files currently put each paragraph in a single line. Would it better to rewrap the paragraphs, or change `max_line_length` to "unset" for Markdown files?

**Comment by faho** [comment]
>Would it better to rewrap the paragraphs, or change max_line_length to "unset" for Markdown files?

Yes. For .md and .txt files, we want wrapping behavior. The ".txt" ones are fed to doxygen, which can do much nicer wrapping when more is on a single line. At least that's what doc_src/FORMATTING.md says. For .md, it's kinda typical to require an empty line between blocks (and soft-wrap everything else), so it's not that important.

Anyway, I'd prefer that in a different PR.

**Comment by faho** [comment]
Merged, thanks!

**Pull request merged**